openpype/hosts/maya/plugins/publish/collect_shots.py

```diff
@@ -0,0 +1,148 @@
+        if grand_parent not in parents:
+            parents.append(grand_parent)
+
+    return parents
+
+
+class CollectShotData(pyblish.api.InstancePlugin):
+    """Collect shot data."""
+
+    # Offset to before CollectHierarchy
+    order = pyblish.api.CollectorOrder - 0.077
+    label = "Collect Shot Data"
+    families = ["shot"]
+
+    def process(self, instance):
+        node = cmds.sets(instance.data["instance_node"], query=True)[0]
+        frame_start = cmds.getAttr(node + ".timeRangeStart")
+        frame_end = cmds.getAttr(node + ".timeRangeStop")
+        asset_entity = instance.context.data["assetEntity"]
+        project_entity = instance.context.data["projectEntity"]
+        width_key = "resolutionWidth"
+        height_key = "resolutionHeight"
+        pixelAspect_key = "pixelAspect"
+        instance.data.update(
+            {
+                "range": [frame_start, frame_end],
+                "heroTrack": True,  # Requirement for CollectHierarchy
+                "handleStart": 0,
+                "handleEnd": 0,
+                "frameStart": frame_start,
+                "frameEnd": frame_end,
+                "clipIn": frame_start,
+                "clipOut": frame_end,
+                "fps": mel.eval("currentTimeUnitToFPS()"),
+                "resolutionWidth": asset_entity.get(
+                    width_key, project_entity.get(width_key, 1920)
+                ),
+                "resolutionHeight": asset_entity.get(
+                    height_key, project_entity.get(height_key, 1080)
+                ),
+                "pixelAspect": asset_entity.get(
+                    pixelAspect_key, project_entity.get(pixelAspect_key, 1)
+                ),
+                "asset": instance.data["name"],
+                # Disabling task fetching in collect_anatomy_instance_data
+                "task": None,
+                # For extract_maya_scene_raw.
+                "exactSetMembersOnly": False
+            }
+        )
+
+        # Set hierarchy parents.
+        instance.data["parents"] = []
+        context_parents = instance.context.data.get("parents")
+        if context_parents is None:
+            parents = get_parents(project_entity["name"], asset_entity)
+            for parent in reversed(parents):
+                instance.data["parents"].append(
+                    {
+                        "entity_name": parent["name"],
+                        "entity_type": parent["type"]
+                    }
+                )
+            instance.data["parents"].append(
+                {
+                    "entity_name": asset_entity["name"],
+                    "entity_type": asset_entity["type"]
+                }
+            )
+            instance.context.data["parents"] = instance.data["parents"]
+        else:
+            instance.data["parents"] = instance.context.data["parents"]
+
+        self.log.debug(
+            json.dumps(instance.data["parents"], indent=4, sort_keys=True)
+        )
```

Comment:
I don't get this - why is the `context` the ground truth for this? I believe the fact that this relies on Context asset entity instead of the instance's asset data is actually bad practice to begin with. I understand the logic for it; because you most likely want it related to current context but that's not how the publishing system is designed. It's totally valid for a shot instance to be targeting a different asset entity; right?

Or what's the thinking behind this?

Also if we really really really want to do this only once - why not put it in the collector below that is actually a ContextPlugin?

Reply:
This was to try and save on database queries. I theory this should only query the data once for the parents, then subsequent instances will use the same parents.

There is no explicit targeting of the parent of the shots. It's implicit that shots will be created under the current asset. So launching `sequence01/layout` task, would create shots like `sequence01/shot01` etc. Not sure how you would be able to target another asset as parent, in the publishing UI?

Reply:
I could actually just do this in the `CollectShotsData` instead?

Reply:
> I could actually just do this in the CollectShotsData instead?

Yes, that's what I meant with my last sentence.

> Not sure how you would be able to target another asset as parent, in the publishing UI?

You could just change the asset (and task) in the publisher UI, right?